## PROJECT CONTEXT
To predict INFY,TCS prices for tomorrow

## PROBLEM SPECIFICS
- Deliverable: Trained model file and bare python file
- Machine learning task: Regression
- Target variable: Stock Price

<font size='6' color='red'>EXPLORATORY ANALYSIS</font>

In [33]:
from datetime import date
from nsepy import get_history
import pandas as pd
import numpy as np
from bokeh.plotting import figure
from bokeh.io import output_notebook, show
output_notebook()

Loading BokehJS ...

In [2]:
infosys_df = get_history(symbol='INFY',
                         start=date(2015,1,1),
                        end=date(2016,12,31))
infosys_df.head()

Symbol Series  Prev Close     Open     High     Low     Last  \
Date                                                                      
2015-01-01   INFY     EQ     1972.55  1968.95  1982.00  1956.9  1971.00   
2015-01-02   INFY     EQ     1974.40  1972.00  2019.05  1972.0  2017.95   
2015-01-05   INFY     EQ     2013.20  2009.90  2030.00  1977.5  1996.00   
2015-01-06   INFY     EQ     1995.90  1980.00  1985.00  1934.1  1965.10   
2015-01-07   INFY     EQ     1954.20  1965.00  1974.75  1950.0  1966.05   

              Close     VWAP   Volume      Turnover  Trades  \
Date                                                          
2015-01-01  1974.40  1971.34   500691  9.870306e+13   14908   
2015-01-02  2013.20  2003.25  1694580  3.394669e+14   54166   
2015-01-05  1995.90  2004.59  2484256  4.979911e+14   82694   
2015-01-06  1954.20  1954.82  2416829  4.724458e+14  108209   
2015-01-07  1963.55  1962.59  1812479  3.557162e+14   62463   

            Deliverable Volume  %Deliverble  
Date                                         
2015-01-01              258080       0.5154  
2015-01-02             1249104       0.7371  
2015-01-05             1830962       0.7370  
2015-01-06             1772070       0.7332  
2015-01-07             1317720       0.7270

In [3]:
tcs_df = get_history(symbol='TCS',
                     start=date(2015,1,1),
                     end=date(2016,12,31))
tcs_df.head()

Symbol Series  Prev Close    Open     High      Low     Last  \
Date                                                                      
2015-01-01    TCS     EQ     2558.25  2567.0  2567.00  2541.00  2550.00   
2015-01-02    TCS     EQ     2545.55  2551.0  2590.95  2550.60  2588.40   
2015-01-05    TCS     EQ     2579.45  2581.0  2599.90  2524.65  2538.10   
2015-01-06    TCS     EQ     2540.25  2529.1  2529.10  2440.00  2450.05   
2015-01-07    TCS     EQ     2446.60  2470.0  2479.15  2407.45  2426.90   

              Close     VWAP   Volume      Turnover  Trades  \
Date                                                          
2015-01-01  2545.55  2548.51   183415  4.674345e+13    8002   
2015-01-02  2579.45  2568.19   462870  1.188740e+14   27585   
2015-01-05  2540.25  2563.94   877121  2.248886e+14   43234   
2015-01-06  2446.60  2466.90  1211892  2.989615e+14   84503   
2015-01-07  2417.70  2433.96  1318166  3.208362e+14  101741   

            Deliverable Volume  %Deliverble  
Date                                         
2015-01-01               52870       0.2883  
2015-01-02              309350       0.6683  
2015-01-05              456728       0.5207  
2015-01-06              714306       0.5894  
2015-01-07              886368       0.6724

In [4]:
nifty_df = get_history(symbol='NIFTY IT',
                     start=date(2015,1,1),
                     end=date(2016,12,31),
                      index=True)
nifty_df.head()

Open      High       Low     Close    Volume      Turnover
Date                                                                      
2015-01-01  11214.80  11235.75  11166.35  11215.70   4246150  3.575100e+09
2015-01-02  11214.65  11399.10  11214.65  11372.10  10004862  9.645600e+09
2015-01-05  11369.35  11433.75  11186.95  11248.55   8858018  1.059000e+10
2015-01-06  11186.10  11186.10  10909.00  10959.90  12515739  1.364500e+10
2015-01-07  11013.20  11042.35  10889.55  10916.00  10976356  1.203440e+10

In [5]:
print(infosys_df.shape, tcs_df.shape, nifty_df.shape)

(495, 14) (495, 14) (495, 6)


In [6]:
infosys_df.dtypes[infosys_df.dtypes!=object]

Prev Close            float64
Open                  float64
High                  float64
Low                   float64
Last                  float64
Close                 float64
VWAP                  float64
Volume                  int64
Turnover              float64
Trades                  int64
Deliverable Volume      int64
%Deliverble           float64
dtype: object

### Moving average function

In [7]:
def mov_avg(df,window):
    mean = df.rolling(window).mean()
    ma = mean
    return ma 

In [8]:
mov_avg(infosys_df['Close'],10)

Date
2015-01-01         NaN
2015-01-02         NaN
2015-01-05         NaN
2015-01-06         NaN
2015-01-07         NaN
2015-01-08         NaN
2015-01-09         NaN
2015-01-12         NaN
2015-01-13         NaN
2015-01-14    2028.265
2015-01-15    2044.250
2015-01-16    2054.760
2015-01-19    2065.420
2015-01-20    2082.420
2015-01-21    2103.055
2015-01-22    2125.355
2015-01-23    2139.415
2015-01-27    2141.470
2015-01-28    2147.035
2015-01-29    2148.720
2015-01-30    2149.485
2015-02-02    2151.340
2015-02-03    2153.180
2015-02-04    2155.045
2015-02-05    2157.435
2015-02-06    2160.840
2015-02-09    2164.225
2015-02-10    2178.405
2015-02-11    2192.435
2015-02-12    2209.005
                ...   
2016-11-21     940.515
2016-11-22     934.140
2016-11-23     927.905
2016-11-24     925.105
2016-11-25     928.915
2016-11-28     934.735
2016-11-29     939.640
2016-11-30     943.105
2016-12-01     947.685
2016-12-02     952.135
2016-12-05     957.190
2016-12-06     962.420
2016-1

In [9]:
type(mov_avg(infosys_df['Close'],10))

pandas.core.series.Series

### Visualising trends of both stock and index between 2015-2016

In [10]:
x=infosys_df.index
y=infosys_df['Close']
z=tcs_df['Close']
a=nifty_df['Close']
p = figure(title='infy stock trends',x_axis_label='Timeline', y_axis_label='Stock closing price',x_axis_type='datetime',height=300)
p.line(x,y,legend='INFY',line_color='red')
p.line(x,z,legend='TCS',line_color='green')
p.line(x,a,legend='NIFTY IT')
show(p)

<font size='6' color='red'>FEATURE ENGINEERING</font>

### CREATING DUMMY TIME SERIES

In [11]:
# VOLUME SHOCKS
infosys_df['Volume_shock']=((abs(infosys_df.Volume-infosys_df.Volume.shift())/infosys_df.Volume)>=0.10).astype(int)
infosys_df.head()

Symbol Series  Prev Close     Open     High     Low     Last  \
Date                                                                      
2015-01-01   INFY     EQ     1972.55  1968.95  1982.00  1956.9  1971.00   
2015-01-02   INFY     EQ     1974.40  1972.00  2019.05  1972.0  2017.95   
2015-01-05   INFY     EQ     2013.20  2009.90  2030.00  1977.5  1996.00   
2015-01-06   INFY     EQ     1995.90  1980.00  1985.00  1934.1  1965.10   
2015-01-07   INFY     EQ     1954.20  1965.00  1974.75  1950.0  1966.05   

              Close     VWAP   Volume      Turnover  Trades  \
Date                                                          
2015-01-01  1974.40  1971.34   500691  9.870306e+13   14908   
2015-01-02  2013.20  2003.25  1694580  3.394669e+14   54166   
2015-01-05  1995.90  2004.59  2484256  4.979911e+14   82694   
2015-01-06  1954.20  1954.82  2416829  4.724458e+14  108209   
2015-01-07  1963.55  1962.59  1812479  3.557162e+14   62463   

            Deliverable Volume  %Deliverble  Volume_shock  
Date                                                       
2015-01-01              258080       0.5154             0  
2015-01-02             1249104       0.7371             1  
2015-01-05             1830962       0.7370             1  
2015-01-06             1772070       0.7332             0  
2015-01-07             1317720       0.7270             1

In [12]:
# PRICE SHOCKS
infosys_df['Price_shock']=(((abs(infosys_df.Close-infosys_df['Prev Close']))/infosys_df.Close)>=0.02).astype(int)
infosys_df.head()

Symbol Series  Prev Close     Open     High     Low     Last  \
Date                                                                      
2015-01-01   INFY     EQ     1972.55  1968.95  1982.00  1956.9  1971.00   
2015-01-02   INFY     EQ     1974.40  1972.00  2019.05  1972.0  2017.95   
2015-01-05   INFY     EQ     2013.20  2009.90  2030.00  1977.5  1996.00   
2015-01-06   INFY     EQ     1995.90  1980.00  1985.00  1934.1  1965.10   
2015-01-07   INFY     EQ     1954.20  1965.00  1974.75  1950.0  1966.05   

              Close     VWAP   Volume      Turnover  Trades  \
Date                                                          
2015-01-01  1974.40  1971.34   500691  9.870306e+13   14908   
2015-01-02  2013.20  2003.25  1694580  3.394669e+14   54166   
2015-01-05  1995.90  2004.59  2484256  4.979911e+14   82694   
2015-01-06  1954.20  1954.82  2416829  4.724458e+14  108209   
2015-01-07  1963.55  1962.59  1812479  3.557162e+14   62463   

            Deliverable Volume  %Deliverble  Volume_shock  Price_shock  
Date                                                                    
2015-01-01              258080       0.5154             0            0  
2015-01-02             1249104       0.7371             1            0  
2015-01-05             1830962       0.7370             1            0  
2015-01-06             1772070       0.7332             0            1  
2015-01-07             1317720       0.7270             1            0

In [13]:
# price shock without volume shock
infosys_df['priceshock_wihout_volshock']=((infosys_df.Price_shock==1) & (infosys_df.Volume_shock==0)).astype(int)
infosys_df.head()

Symbol Series  Prev Close     Open     High     Low     Last  \
Date                                                                      
2015-01-01   INFY     EQ     1972.55  1968.95  1982.00  1956.9  1971.00   
2015-01-02   INFY     EQ     1974.40  1972.00  2019.05  1972.0  2017.95   
2015-01-05   INFY     EQ     2013.20  2009.90  2030.00  1977.5  1996.00   
2015-01-06   INFY     EQ     1995.90  1980.00  1985.00  1934.1  1965.10   
2015-01-07   INFY     EQ     1954.20  1965.00  1974.75  1950.0  1966.05   

              Close     VWAP   Volume      Turnover  Trades  \
Date                                                          
2015-01-01  1974.40  1971.34   500691  9.870306e+13   14908   
2015-01-02  2013.20  2003.25  1694580  3.394669e+14   54166   
2015-01-05  1995.90  2004.59  2484256  4.979911e+14   82694   
2015-01-06  1954.20  1954.82  2416829  4.724458e+14  108209   
2015-01-07  1963.55  1962.59  1812479  3.557162e+14   62463   

            Deliverable Volume  %Deliverble  Volume_shock  Price_shock  \
Date                                                                     
2015-01-01              258080       0.5154             0            0   
2015-01-02             1249104       0.7371             1            0   
2015-01-05             1830962       0.7370             1            0   
2015-01-06             1772070       0.7332             0            1   
2015-01-07             1317720       0.7270             1            0   

            priceshock_wihout_volshock  
Date                                    
2015-01-01                           0  
2015-01-02                           0  
2015-01-05                           0  
2015-01-06                           1  
2015-01-07                           0

## Handling unequal time series

In [112]:
idx = pd.date_range('1-1-2015','30-12-2016')
idx

DatetimeIndex(['2015-01-01', '2015-01-02', '2015-01-03', '2015-01-04',
               '2015-01-05', '2015-01-06', '2015-01-07', '2015-01-08',
               '2015-01-09', '2015-01-10',
               ...
               '2016-12-21', '2016-12-22', '2016-12-23', '2016-12-24',
               '2016-12-25', '2016-12-26', '2016-12-27', '2016-12-28',
               '2016-12-29', '2016-12-30'],
              dtype='datetime64[ns]', length=730, freq='D')

In [113]:
aug_infydf = infosys_df.reindex(idx)
print(infosys_df.shape)
aug_infydf.isnull().sum()

(495, 17)


Symbol                        235
Series                        235
Prev Close                    235
Open                          235
High                          235
Low                           235
Last                          235
Close                         235
VWAP                          235
Volume                        235
Turnover                      235
Trades                        235
Deliverable Volume            235
%Deliverble                   235
Volume_shock                  235
Price_shock                   235
priceshock_wihout_volshock    235
dtype: int64

#### As we can see there are 235 holidays in between 2015 to 2016 out of total 730 days i.e. around 32% of observations is missing. Hence, filling out these values might not be a good strategy and we might end up adding our own biases to the dataset.
##### We will do feature engineering by adding a new indicator variable named 'holiday' which will tell our model about the missing data due to holidays.

In [114]:
# adding new feature which indicates if a date is holiday
aug_infydf['Holiday']=aug_infydf['Close'].isnull().astype(int)

In [115]:
# as sklearn cannot handle Nan, hence we will fill missing values of holidays with 0
aug_infydf.fillna(0,inplace=True)
aug_infydf.head()

Symbol Series  Prev Close     Open     High     Low     Last  \
2015-01-01   INFY     EQ     1972.55  1968.95  1982.00  1956.9  1971.00   
2015-01-02   INFY     EQ     1974.40  1972.00  2019.05  1972.0  2017.95   
2015-01-03      0      0        0.00     0.00     0.00     0.0     0.00   
2015-01-04      0      0        0.00     0.00     0.00     0.0     0.00   
2015-01-05   INFY     EQ     2013.20  2009.90  2030.00  1977.5  1996.00   

             Close     VWAP     Volume      Turnover   Trades  \
2015-01-01  1974.4  1971.34   500691.0  9.870306e+13  14908.0   
2015-01-02  2013.2  2003.25  1694580.0  3.394669e+14  54166.0   
2015-01-03     0.0     0.00        0.0  0.000000e+00      0.0   
2015-01-04     0.0     0.00        0.0  0.000000e+00      0.0   
2015-01-05  1995.9  2004.59  2484256.0  4.979911e+14  82694.0   

            Deliverable Volume  %Deliverble  Volume_shock  Price_shock  \
2015-01-01            258080.0       0.5154           0.0          0.0   
2015-01-02           1249104.0       0.7371           1.0          0.0   
2015-01-03                 0.0       0.0000           0.0          0.0   
2015-01-04                 0.0       0.0000           0.0          0.0   
2015-01-05           1830962.0       0.7370           1.0          0.0   

            priceshock_wihout_volshock  Holiday  
2015-01-01                         0.0        0  
2015-01-02                         0.0        0  
2015-01-03                         0.0        1  
2015-01-04                         0.0        1  
2015-01-05                         0.0        0

<font size='4' color='red'>DATA VISUALIZATION</font>

### For INFY

In [72]:
a = mov_avg(infosys_df['Close'],52)

In [103]:

x=infosys_df.index
y=infosys_df['Close']
df1 = infosys_df[ infosys_df.Volume_shock==1]
df2 = infosys_df[infosys_df.priceshock_wihout_volshock==1]
x1=[b for b in df1.index]
y1=[a for a in df1['Close']]
x2=[b for b in df2.index]
y2=[a for a in df2['Close']]
p = figure(title='infy stock trends',x_axis_label='Timeline', y_axis_label='Stock closing price',x_axis_type='datetime',width=1000)
p.line(x,y,legend='INFY closing price',line_width=2)
p.diamond(x1,y1,fill_color='red',line_color='red',legend='Volume shock')
p.line(x,a,color='orange',legend='52 week mov.avg')
p.circle(x2,y2,fill_color='black',line_color='black',legend='Volumeless price movement')

#p.patch
show(p)

## ABT Construction

In [116]:
aug_infydf.head()

Symbol Series  Prev Close     Open     High     Low     Last  \
2015-01-01   INFY     EQ     1972.55  1968.95  1982.00  1956.9  1971.00   
2015-01-02   INFY     EQ     1974.40  1972.00  2019.05  1972.0  2017.95   
2015-01-03      0      0        0.00     0.00     0.00     0.0     0.00   
2015-01-04      0      0        0.00     0.00     0.00     0.0     0.00   
2015-01-05   INFY     EQ     2013.20  2009.90  2030.00  1977.5  1996.00   

             Close     VWAP     Volume      Turnover   Trades  \
2015-01-01  1974.4  1971.34   500691.0  9.870306e+13  14908.0   
2015-01-02  2013.2  2003.25  1694580.0  3.394669e+14  54166.0   
2015-01-03     0.0     0.00        0.0  0.000000e+00      0.0   
2015-01-04     0.0     0.00        0.0  0.000000e+00      0.0   
2015-01-05  1995.9  2004.59  2484256.0  4.979911e+14  82694.0   

            Deliverable Volume  %Deliverble  Volume_shock  Price_shock  \
2015-01-01            258080.0       0.5154           0.0          0.0   
2015-01-02           1249104.0       0.7371           1.0          0.0   
2015-01-03                 0.0       0.0000           0.0          0.0   
2015-01-04                 0.0       0.0000           0.0          0.0   
2015-01-05           1830962.0       0.7370           1.0          0.0   

            priceshock_wihout_volshock  Holiday  
2015-01-01                         0.0        0  
2015-01-02                         0.0        0  
2015-01-03                         0.0        1  
2015-01-04                         0.0        1  
2015-01-05                         0.0        0

In [117]:
# DROPPING IRRELEVANT AND REDUNDANT FEATURES
# Symbol and Series are same for each observation and Prev Close is redundant to Close
aug_infydf = aug_infydf.drop(['Symbol','Series','Prev Close'],axis=1)
aug_infydf.head()

Open     High     Low     Last   Close     VWAP     Volume  \
2015-01-01  1968.95  1982.00  1956.9  1971.00  1974.4  1971.34   500691.0   
2015-01-02  1972.00  2019.05  1972.0  2017.95  2013.2  2003.25  1694580.0   
2015-01-03     0.00     0.00     0.0     0.00     0.0     0.00        0.0   
2015-01-04     0.00     0.00     0.0     0.00     0.0     0.00        0.0   
2015-01-05  2009.90  2030.00  1977.5  1996.00  1995.9  2004.59  2484256.0   

                Turnover   Trades  Deliverable Volume  %Deliverble  \
2015-01-01  9.870306e+13  14908.0            258080.0       0.5154   
2015-01-02  3.394669e+14  54166.0           1249104.0       0.7371   
2015-01-03  0.000000e+00      0.0                 0.0       0.0000   
2015-01-04  0.000000e+00      0.0                 0.0       0.0000   
2015-01-05  4.979911e+14  82694.0           1830962.0       0.7370   

            Volume_shock  Price_shock  priceshock_wihout_volshock  Holiday  
2015-01-01           0.0          0.0                         0.0        0  
2015-01-02           1.0          0.0                         0.0        0  
2015-01-03           0.0          0.0                         0.0        1  
2015-01-04           0.0          0.0                         0.0        1  
2015-01-05           1.0          0.0                         0.0        0

In [118]:
aug_infydf.tail()

Open    High     Low     Last    Close     VWAP     Volume  \
2016-12-26   988.00   994.0  975.10   983.35   982.75   983.28  1606914.0   
2016-12-27   982.75  1001.9  980.00   996.90   998.95   993.09  2456408.0   
2016-12-28  1002.95  1009.8  995.00   999.00   998.50  1003.15  2738408.0   
2016-12-29  1003.75  1008.0  988.15   990.00   992.35   997.88  3120062.0   
2016-12-30   998.00  1012.0  992.75  1011.00  1010.60  1006.91  3633884.0   

                Turnover    Trades  Deliverable Volume  %Deliverble  \
2016-12-26  1.580050e+14   47565.0            958338.0       0.5964   
2016-12-27  2.439426e+14   57508.0           1430604.0       0.5824   
2016-12-28  2.747040e+14   81224.0           1853359.0       0.6768   
2016-12-29  3.113452e+14   92963.0           2303901.0       0.7384   
2016-12-30  3.659012e+14  100569.0           2574459.0       0.7085   

            Volume_shock  Price_shock  priceshock_wihout_volshock  Holiday  
2016-12-26           1.0          0.0                         0.0        0  
2016-12-27           1.0          0.0                         0.0        0  
2016-12-28           1.0          0.0                         0.0        0  
2016-12-29           1.0          0.0                         0.0        0  
2016-12-30           1.0          0.0                         0.0        0

In [119]:
aug_infydf.shape

(730, 15)

In [121]:
# finding autocorrelation of all the features
for feature in aug_infydf:
    print(aug_infydf[feature].autocorr())

0.396816694827433
0.39590718569154215
0.3969085583346883
0.39661206985934694
0.39637917672617806
0.3962388476883521
0.20426853056346753
0.18748217722949048
0.23044046882935362
0.23058386465071212
0.2827068381415425
0.11017484522541397
0.08254590468330907
-0.011095700416088787
0.2715651649582232


 We can see that there is no strong autocorrelation in any of the features
 ### Saving the augmented engineered data as csv

In [120]:
aug_infydf.to_csv('analytical_base_table.csv',index=)